Add support for Japanese GPT-NeoX-based model by ABEJA, Inc.

SO0529 opened the pull request:
# What does this PR do?
This PR adds a new GPT NeoX Japanese Model and a new Tokenizer. The specific features are,
- Trained for Japanese Dataset with specific preprocess.
- Used [GPT-NeoX](https://github.com/EleutherAI/gpt-neox) to train with Pipe. In addition, we removed bias parameters from the transformer blocks following [PaLM from Google](https://ai.googleblog.com/2022/04/pathways-language-model-palm-scaling-to.html).
- Applied Japanese special sub-word tokenizer to accommodate the distinctive structure of the Japanese. Japanese has a relatively large vocabulary and there is no separation between words. Furthermore, the language is a combination of hiragana, katakana, and kanji, and variants such as "1" and "①" are often used. 

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Thank you in advance to review this PR!
- GPT model and tokenizer: @patrickvonplaten, @LysandreJik
- Documentation: @sgugger


<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

SO0529 committed:
add gpt-neox-japanese model and tokenizer as new model

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18814). All of your documentation changes will be reflected on that endpoint.

LysandreJik commented:
Impressive PR, @SO0529!

I'd like either @younesbelkada or @ArthurZucker (or both :smile:) to have a look at your PR; they are both well versed in Japanese and have reviewed models in the past.

They're both on leave until the 9th of September, is it okay with you if we review this PR in about 1.5 weeks' time?

Thanks for your understanding!

SO0529 commented:
@LysandreJik 
Thank you for a quick comment. Of course we can wait! We look forward to having you review this PR!

ArthurZucker commented:
Hey @SO0529 ! We just came back, gonna take a look at this with @younesbelkada asap 😄

SO0529 commented:
@ArthurZucker @younesbelkada Thank you in advance for taking your time to review this PR!:smiley:

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,788 @@
+
+class RotaryEmbedding(torch.nn.Module):
+    """
+    same class as original gpt-neox
```
Here you could just copy the entire class from `GPT-NeoX` and add a line on top of the class definition that states 
`# Coped from ....`
Check for instance this [line](https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/src/transformers/models/opt/modeling_opt.py#L523) from OPT

SO0529 replied:
I've modified following your comment!

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,788 @@
+        # Need to be a tensor, otherwise we get error: `RuntimeError: expected scalar type float but found double`.
+        # Need to be on the same device, otherwise `RuntimeError: ..., x and y to be on the same device`
+        mask_value = torch.tensor(mask_value, dtype=attn_scores.dtype).to(attn_scores.device)
+        attn_scores = torch.where(causal_mask, attn_scores, mask_value)
```
When I tried to run slow tests on my GPU, I got an error that I fixed by adding the following line! Could you please double check that you get the same error by making sure `accelerate` is installed in your virtual environment 🙏 
```suggestion
        causal_mask = causal_mask.to(attn_scores.device)
        attn_scores = torch.where(causal_mask, attn_scores, mask_value)
```

younesbelkada replied:
Actually I suspect this might be fixed by avoiding using `F.linear`

SO0529 replied:
I've modified following your comment! As you pointed out, this was happening because the devices in the masks were different.

younesbelkada replied:
Perfect thanks a lot!

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,788 @@
+
+
+class RowParallelLinear(nn.Module):
+    def __init__(self, input_size, output_size, bias=False) -> None:
```
I can see that these classes are not implemented in the GPT-NeoX implementation, I think that it would be great to add a bit more documentation on why we need to implement a custom `RowParallelLinear`, I suspect you did used `Megatron-DeepSpeed` to train your models? 
Also I suspect that you can replace this module by a simple linear module but would involve a bit of debugging I guess 💪

SO0529 replied:
Exactly we used `Megatron-DeepSpeed` and this name of class came from there. I removed all `RowParallelLinear` and `ColumnParallelLinear`.
Currently this code passes cpu env, but gave me an error regarding in [def bias_dropout_add](https://github.com/SO0529/transformers/blob/gpt-neox-japanese/src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py#L292) under gpu env.
The test error log is like below.
```
    def bias_dropout_add(x: Tensor, bias: Tensor, residual: Optional[Tensor], prob: float, training: bool) -> Tensor:
        if bias is not None:
>           x = x + bias
E           RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and meta!
```
I just changed like `x = x + bias.to(x.device)`, but same results... I need to deep dive about it.

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,788 @@
+
+
+class ColumnParallelLinear(torch.nn.Module):
+    def __init__(self, input_size, output_size, bias=False) -> None:
```
Here I have to same comments as above! 🔥

SO0529 replied:
I removed `ColumnParallelLinear` same as `RowParallelLinear`:ok_hand:

younesbelkada commented on tests/models/gpt_neox_japanese/test_modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,233 @@
+
+    @unittest.skip(reason="Feed forward chunking is not implemented")
+    def test_feed_forward_chunking(self):
+        pass
```
Do you think that we can add some slow generation and/or logits tests with a small model? You can check some nice tests here: https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/tests/models/opt/test_modeling_opt.py#L363 !

SO0529 replied:
I add [test_generation](https://github.com/SO0529/transformers/blob/gpt-neox-japanese/tests/models/gpt_neox_japanese/test_modeling_gpt_neox_japanese.py#L232) with slow.

younesbelkada reviewed (changes requested):
Hi @SO0529 !
First of all I would like to congratulate on the impressive results you have obtained with this model and this amazing PR! It is a pleasure to see amazing contributions from the Japanese community
 
I went through a quick review and left several questions that might need to be addressed before merging but overall I think that:
- We might need a bit more documentation about `RowParrallelLinear` as it is a new module that is not present in the original GPT-NeoX implementation. I also think that the name can be changed as I think that there is no model parallelism involved (we just do a `F.linear`). Feel free to have a look at the BLOOM implementation to see [how we got rid of the `output_bias` variable. ](https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/src/transformers/models/bloom/modeling_bloom.py#L383) there. But I am not sure yet how to adapt this with your model since the argument `skip_bias_add` might be important (for BLOOM this argument was set to `False` on all models).
-  A small nit on the Attention block that can be easily fixed with the small change that I proposed, but I feel that the error initially came from using `F.linear` on the `RowParrallelLinear` and `ColumnParrallelLinear` modules (`accelerate` does not support `torch.functionnal` functions so that is why it might be related).
- If some modules are entirely copied from the original `GPT-NeoX` implementation, you can just add a `# Copied from...` statement on the top of the class definition for better tracking 💪 
- Also not sure if we have to keep the `bias_dropout_fn` as I experienced a very small throughput enhancement that we can neglect for better code readability. Feel free to have a look [https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/src/transformers/models/bloom/modeling_bloom.py#L130](here) as we went through the same dilemma when integrating BLOOM from Megatron-DeepSpeed

Great effort on the integration side and looking forward to reviewing your next changes and discussing with you as I am very curious to understand the challenges you have faced to integrate the architecture!

younesbelkada reviewed (commented):
Commenting here to try to change the status of my comment to "Comment" instead of "Request changes"

younesbelkada reviewed (commented):
Commenting here to try to change the status of my comment to "Comment" instead of "Request changes"

SO0529 commented:
Thanks for your valuable comments! I correct them one by one:muscle:

ArthurZucker commented:
Awesome, I will review again once you are done with that 😉

SO0529 committed:
Correction to PR's comment for GPT NeoX Japanese
- Fix to be able to use gpu
- Add comment # Copied... at the top of RotaryEmbedding
- Implement nn.Linear instead of original linear class
- Add generation test under @slow

SO0529 commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,788 @@
+            init.uniform_(self.bias, -bound, bound)
+
+    def forward(self, input_):
+        output = F.linear(input_, self.weight)
```
I removed all `F.linear` from this model:smiley:

SO0529 commented:
Sorry I found mistakes, I re-push later when I fix.

SO0529 committed:
fix bias treatment for gpt-neox-japanese

SO0529 commented:
@younesbelkada @ArthurZucker 
I think I could fix all your comments like below. I'm glad if you could review this PR again:smile:

> - We might need a bit more documentation about RowParrallelLinear as it is a new module that is not present in the original GPT-NeoX implementation. I also think that the name can be changed as I think that there is no model parallelism involved (we just do a F.linear). Feel free to have a look at the BLOOM implementation to see [how we got rid of the output_bias variable. ](https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/src/transformers/models/bloom/modeling_bloom.py#L383)there. But I am not sure yet how to adapt this with your model since the argument skip_bias_add might be important (for BLOOM this argument was set to False on all models).
- I removed both `RowParrallelLinear` and `ColumnParrallelLinear` from this model.

> - A small nit on the Attention block that can be easily fixed with the small change that I proposed, but I feel that the error initially came from using F.linear on the RowParrallelLinear and ColumnParrallelLinear modules (accelerate does not support torch.functionnal functions so that is why it might be related).
- I used `nn.Linear` instead of `F.linear`.

> - If some modules are entirely copied from the original GPT-NeoX implementation, you can just add a # Copied from... statement on the top of the class definition for better tracking 💪
- I put a comment at the top of class definition.

> - Also not sure if we have to keep the bias_dropout_fn as I experienced a very small throughput enhancement that we can neglect for better code readability. Feel free to have a look [https://github.com/huggingface/transformers/blob/9faa9f9dacf8c818ab2513da3ef92ce66f39515d/src/transformers/models/bloom/modeling_bloom.py#L130](https://github.com/huggingface/transformers/pull/here) as we went through the same dilemma when integrating BLOOM from Megatron-DeepSpeed
- I keep `bias_dropout_add` function to use the bias param, but it has been changed from an unnecessarily difficult-to-read writing style to a simple description.

In addition to above, I added a simple `test_generate` function.

Again, thank you for taking your time to review this PR!

SO0529 commented:
By the way, `build_and_test ` and `ci/circleci: run_tests_torch` looks to be in error with the model `tests/models/pegasus/test_modeling_pegasus.py`, how do I address this error?:worried:

ArthurZucker commented:
Hey! Awesome work, we will review again today! 
The Pegasus test should not really be related to you, I will have a look 😊

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,719 @@
+        return attn_output, attn_weights
+
+
+# Copied from transformers.models.gpt_neox.modeling_gpt_neox.RotaryEmbedding
```
Very nice 💪 !!

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,719 @@
+    return q_embed, k_embed
+
+
+def bias_dropout_add(x: Tensor, bias: Tensor, residual: Optional[Tensor], prob: float, training: bool) -> Tensor:
```
A bit more documentation here would be great ;)

SO0529 replied:
I wrote a documentation like below.
```
"""add bias to x, apply dropout and residual connection

    Args:
        x (Tensor): main path of output
        bias (Tensor): None or attn_bias of the last attention layer
        residual (Optional[Tensor]): residual value
        prob (float): dropout probability
        training (bool): whether in training mode or not

    Returns:
        Tensor: dropout(x + bias) + residual
"""
```

younesbelkada commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,719 @@
+        self.config = config
+
+        self.embed_in = nn.Embedding(config.vocab_size, config.hidden_size)
+        self.layers = nn.ModuleList([])
+        for layer_i in range(config.num_hidden_layers):
+            self.layers.append(GPTNeoXJapaneseLayer(config=config, layer_number=layer_i))
```
I suggest to change it to 👇 
```suggestion
        self.layers = nn.ModuleList([(GPTNeoXJapaneseLayer(config=config, layer_number=i) for i in range(config.num_hidden_layers)])
```
But feel free to ignore my suggestion (also you might need to run `make style` after accepting the suggestion)

SO0529 replied:
I modified here and ran `make style` following your comment:smile:

younesbelkada commented on tests/models/gpt_neox_japanese/test_modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,255 @@
+
+    @slow
+    def test_generation(self):
+        model_id = "abeja/gpt-neox-japanese-2.7b"
```
Great test ! 🔥 
We might need to use a smaller model probably, even if it is not fully trained that's fine.
If we don't have any other choices I think that it is still fine

SO0529 replied:
Sorry 2.7B model is the minimum model we trained...

younesbelkada reviewed (approved):
Thank you very much @SO0529 for iterating very quickly on the feedback 🔥 
Looking very good now as you have handled very well the conversion between Megatron-DeepSpeed and `transformers` - regarding using `F.Linear`, we advocate to use simpler solutions (`nn.Linear`) for better code readability. 
Looking also very good on the test side, we might ask you to use a smaller model (in the generation test you used the `2B` parameters model that might take long to download on our CI machines)
Will approve for now to remove the "Request changes" and let @ArthurZucker take over the rest of the PR

Again thanks a lot for your great work! 
よろしくお願いします!

SO0529 committed:
Modidy gpt-neox-japanese following PR
- add doc for bias_dropout_add
- style change following a PR comment

SO0529 commented:
@younesbelkada 
Thank you for quick review!
I modified below 2 items following your comments.
- I put a documentation for `bias_dropout_add`.
- I changed coding style regarding `nn.ModuleList`.

sgugger commented on docs/source/en/_toctree.yml:
```diff
@@ -253,6 +253,8 @@
         title: GPT Neo
       - local: model_doc/gpt_neox
         title: GPT NeoX
+      - local: model_doc/gpt_neox_japanese
```
You should also add the model to the main README.

SO0529 replied:
I wrote README, README_ko, README_zh-hans and README_zh-hant.

sgugger commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,731 @@
+
+        # Attention mask.
+        if attention_mask is not None:
+            assert batch_size > 0, "batch_size has to be defined and > 0"
```
No new asserts in the code base please. Use a test and raise the appropriate error!

SO0529 replied:
I used `raise` instead of asset:muscle:

sgugger commented on src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,731 @@
+        self.embed_in = value
+
+    @add_start_docstrings_to_model_forward(GPT_NEOX_JAPANESE_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
+    @add_code_sample_docstrings(
```
You should use `replace_return_docstrings` here as you are rewriting the example. See how to do this by looking at [this model](https://github.com/huggingface/transformers/blob/2886f7f08ac2cd9c0255e3b66d4ee95884b96087/src/transformers/models/bert/modeling_bert.py#L1074) but basically you just need to add en empty Returns to the docstring.

SO0529 replied:
Thank you for giving the sample. I corrected.

sgugger commented on src/transformers/models/gpt_neox_japanese/tokenization_gpt_neox_japanese.py:
````diff
@@ -0,0 +1,375 @@
+    - Conversion of heterographs to the same token_id
+    - Emoji and Emoticon are grouped into 12 types as special tags.
+
+    ```
````
```suggestion
    Example:

    ```python
```

SO0529 replied:
I corrected:smiley:

sgugger commented on src/transformers/models/gpt_neox_japanese/tokenization_gpt_neox_japanese.py:
```diff
@@ -0,0 +1,375 @@
+            File containing the vocabulary.
+        emoji_file (`str`):
+            File containing the emoji.
+        unk_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The unknown token. A token that is not in the vocabulary cannot be converted to an ID and is set to be this
+            token instead.
+        pad_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The token used for padding
+        bos_token (`str`, *optional*, defaults to `<|startoftext|>`):
+            The beginning of sequence token.
+        eos_token (`str`, *optional*, defaults to `<|endoftext|>`):
```
The defaults are all missing a pair of `"` :-)

SO0529 replied:
I added a pair of `"` for all default strings.

sgugger commented on utils/check_copies.py:
```diff
@@ -492,6 +492,7 @@ def check_model_list_copy(overwrite=False, max_per_line=119):
     "Speech2Text2",
     "Vision Encoder decoder",
     "VisionTextDualEncoder",
+    "GPT NeoX Japanese",
```
I think the model should be in the README and this line removed.

SO0529 replied:
I removed here.

sgugger reviewed (approved):
Thanks a lot for adding this new model! I left a few comments.
Also you should add the tokenizer in the AutoTokenizer mapping!

SO0529 committed:
add document for gpt-neox-japanese

SO0529 committed:
remove unused import from gpt-neox-japanese

SO0529 commented:
@sgugger 
Thank you for taking your time to review PR!
I think I could modify all your comments. I would like you to look the correction please.

sgugger commented on README.md:
```diff
@@ -304,6 +304,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[GPT](https://huggingface.co/docs/transformers/model_doc/openai-gpt)** (from OpenAI) released with the paper [Improving Language Understanding by Generative Pre-Training](https://blog.openai.com/language-unsupervised/) by Alec Radford, Karthik Narasimhan, Tim Salimans and Ilya Sutskever.
 1. **[GPT Neo](https://huggingface.co/docs/transformers/model_doc/gpt_neo)** (from EleutherAI) released in the repository [EleutherAI/gpt-neo](https://github.com/EleutherAI/gpt-neo) by Sid Black, Stella Biderman, Leo Gao, Phil Wang and Connor Leahy.
 1. **[GPT NeoX](https://huggingface.co/docs/transformers/model_doc/gpt_neox)** (from EleutherAI) released with the paper [GPT-NeoX-20B: An Open-Source Autoregressive Language Model](https://arxiv.org/abs/2204.06745) by Sid Black, Stella Biderman, Eric Hallahan, Quentin Anthony, Leo Gao, Laurence Golding, Horace He, Connor Leahy, Kyle McDonell, Jason Phang, Michael Pieler, USVSN Sai Prashanth, Shi ...
+1. **[GPT NeoX Japanese](https://huggingface.co/docs/transformers/model_doc/gpt_neox_japanese)** (from ABEJA) released by Shinya Otani, Takayoshi Makabe, Anuj Arora, and Kyo Hattori.
```
```suggestion
1. **[GPT NeoX Japanese](https://huggingface.co/docs/transformers/main/model_doc/gpt_neox_japanese)** (from ABEJA) released by Shinya Otani, Takayoshi Makabe, Anuj Arora, and Kyo Hattori.
```
The model will only be in the main branch after the merge, so we need to link to the main doc. You will need to run `make fix-copies` after applying this suggestion to update the localized READMEs.

SO0529 replied:
I changed the link and ran `make fix-copies` to change other README:smile:

sgugger reviewed (approved):
Thanks for iterating! One last comment and we should be good to merge :-)

SO0529 committed:
fix README for gpt-neox-japanese

SO0529 commented:
@sgugger 
Thank you for taking your time again and again! I've changed your last comment:fire:
We are looking forward to being merge!
